sig-cli/charter.md

```diff
@@ -5,25 +5,81 @@ the Roles and Organization Management outlined in [sig-governance].
+  - Version - e.g. `version`
+  - OpenAPI - e.g. `explain`
+- Pre and Post processing API Resource Config, API Requests and API Responses
+  - `apply` - creates a patch from Resource Config by performing a 3-way diff
+  - `get` - formats API responses into columns
+- Aggregating multiple API Responses and post processing them
+  - `describe` - performs multiple calls and aggregates results
+  - `run --expose` - creates multiple resources
+- Collapsing multiple manual steps into a command
+  - `edit` - collapses `get`, edit a file, `patch` into a single command
+  - `cp` - collapses manual steps for copy a file out of a container
+- Generating Kubernetes Resource Config locally or creating Resources remotely
+  - From commands + arguments + flags - e.g. `create configmap`, `run`, `expose`
+  - From declarative files - Kustomization `configMapGenerator`, `secretMapGenerator`
+- Transforming Kubernetes Resource Config locally or patching remotely
+  - From commands + arguments + flags - e.g. `annotate`, `set image`, `patch`, etc
+  - From declarative files - e.g. Kustomization `commonAnnotations`, `images`, `patches`
+  - From flags - e.g. `-n` - set the namespace of Resources
+- Blocking on propagation of an event or change to the cluster
+  - e.g. `wait`, `rollout`
+- Defining or referencing a collection of Resource Config
+  - e.g. `-f` can reference a url, a file containing multiple Resources, or a directory
+  - e.g. `-R` can traverse a directory recursively
+  - e.g. Kustomization `bases` can refer to other Kustomizations
+  - e.g. Kustomization `resources` can refer to Resource Config files
+- Configure how to talk to a specific cluster from the cli
+  - e.g. `config`
+  - e.g. `--context`
+- Selecting which API group/version to invoke if ambiguous in the context of the command
+  - e.g. `get`, `describe`
+  
+Notably, CLI Tooling also performs operations (generation, transformation, validation)
+that work against collections of Resources or Resource Config - e.g. updating
+things that reference names of resources if transforming the names of resources.
 
 ### Out of scope
 
-SIG CLI is not responsible for command-line tools built and maintained by other
-SIGs, such as kubeadm, which is owned by SIG Cluster Lifecycle. SIG CLI is not
-responsible for defining the Kubernetes API that it interfaces with. The
-Kubernetes API is the responsibility of SIG API Machinery.
+- SIG CLI is not responsible for tools developed outside of the
+  SIG (even if they are part of the broader Kubernetes project).
+- SIG CLI is not responsible for kubectl subcommands developed outside of the
+  SIG (even if they are developed through kubectl extension mechanisms).
+- SIG CLI is not responsible for defining the Kubernetes API or Resource Types
+  that it interfaces with (which is owned by SIG Apps and SIG API Machinery).
+- SIG CLI is not responsible for integrating with *specific tools* or APIs
+  developed outside of the Kubernetes project.
```

Comment:
Can we add application management as something that is out of scope? There are two reasons for this...

1. Things with this level of functionality should be server side. This is even called out in the k8s scope doc Brian started. That way other tools and even things in other languages can use them. They are part of the REST API.
2. In SIG Apps we are working on the Application CRD/Controller for this. It's already scoped and being worked in some capacity using the existing server side direction

Reply:
We all need to work together on this project. SIG CLI is one of the project's horizontals. It supports commands specific to scheduling, autoscaling, node, API machinery, and workloads, for example. SIG Apps should work together with SIG CLI to ensure kubectl supports what is needed in that domain.

Reply:
Do you have a link to a formal definition of what qualifies as application management for this context?  The way I have used the term in the past, many of the existing commands would fall into that category.

Additionally, in the updated charter I have written that kubectl is a proving ground for new API features.  I could imagine a development path where some Application Management functionality could be started in the client and be moved to the server as has been done for other functionality.  (not suggesting that we do this, but I don't know what the plans for AM are or what the owners of the area might want)

Reply:
There are a several reasons I wouldn't want to put something in kubectl for application management right now.

1. The App Def WG figured out that interoperability was a high priority and it's something SIG Apps adopted to support the various tools. Tool authors in this space wanted interoperability. It is now in the SIG Apps charter. If something is in kubectl it's not very interoperable. For example, you can't use it in Dashboard. Applications deployed in one tool should be viewable in another tool and even deleted via a 3rd tool.
2. Work has already started on app management via a CRD/Controller (leveraging kubebuilder). We started at the API layer based on a start from @ant31.
3. The more we use CRDs and controllers the more we learn what is needed to move them to beta and GA. This is a good place to dog food. kubectl being better with these, and not just app management, would benefit more people.
4. Do we have a traceable need for a significant number of users?

Reply:
`kubectl rollout` might be considered "application management". OTOH, like the CRUD commands, apply and kustomize are general-purpose and applicable to all resource types, so are not.

Anyway, as mentioned elsewhere, kubectl can configure autoscaling, cordon nodes, expose services, launch and attach to running pods, and interact with all other resource types in the system -- it's a horizontal. So that will include applications, too, to the degree that concept exists in Kubernetes.

Reply:
> w.r.t. proving new API features, I agree that kubectl shouldn't be proving out new specific Resource APIs. Things like server-side apply and dry-run couldn't have been done as CRDs today so I think we should leave the door open for those types of things. A sig apps example would be kubectl rollout status which cuts across the workloads APIs.

If something like `kubectl rollout status` were implemented today I'd want it to be as a plugin or external tool. Maybe even more than one during experimentation to see what works for what cases.

To continue the example, what if someone did it the way kubectl does it today and someone else did one based on labels where you could see a rollout of a deployment and statefulset that were part of the same thing being rolled out at the same time?

Sometimes it's useful to do multiple experiments, see how they fair against each other, and see what useful bits come.

I'm not criticizing `kubectl rollout status` or the process it went though. I'm just looking at it as an example of how I would approach things today.

Wouldn't it be useful to do this kind of thing as a plugin(s) and they see how it fairs in usefulness (usability + utility) first?

I agree we need to have cross SIG engagement on these forms of issues.